src/backend/elfobj.c

```diff
@@ -1653,17 +1663,18 @@ void obj_tlssections()
     IDXSTR namidx;
     int align = I64 ? 16 : 4;
 
-    int sec = ElfObj::getsegment(".tdata", NULL, SHT_PROGBITS, SHF_ALLOC|SHF_WRITE|SHF_TLS, align);
+    int sec = ElfObj::getsegment(".tdata", NULL, SHT_PROGBITS, tls_section_flags(), align);
     Obj::bytes(sec, 0, align, NULL);
 
     namidx = Obj::addstr(symtab_strings,"_tlsstart");
-    elf_addsym(namidx, 0, align, STT_TLS, STB_GLOBAL, MAP_SEG2SECIDX(sec));
+    const int tls_symbol = (config.flags3 & CFG3android) ? STT_OBJECT : STT_TLS;
+    elf_addsym(namidx, 0, align, tls_symbol, STB_GLOBAL, MAP_SEG2SECIDX(sec));
 
-    ElfObj::getsegment(".tdata.", NULL, SHT_PROGBITS, SHF_ALLOC|SHF_WRITE|SHF_TLS, align);
+    ElfObj::getsegment(".tdata.", NULL, SHT_PROGBITS, tls_section_flags(), align);
 
-    sec = ElfObj::getsegment(".tcommon", NULL, SHT_NOBITS, SHF_ALLOC|SHF_WRITE|SHF_TLS, align);
+    sec = ElfObj::getsegment(".tcommon", NULL, SHT_NOBITS, tls_section_flags(), align);
     namidx = Obj::addstr(symtab_strings,"_tlsend");
-    elf_addsym(namidx, 0, align, STT_TLS, STB_GLOBAL, MAP_SEG2SECIDX(sec));
+    elf_addsym(namidx, 0, align, tls_symbol, STB_GLOBAL, MAP_SEG2SECIDX(sec));
```

Comment:
[You said this before](https://github.com/D-Programming-Language/dmd/pull/3643#discussion_r18504021), but I pointed out that I saw no indication it's true.  Perhaps I'm not following bionic development as closely as you: can you point to why you think TLS is or will be supported?

When I ran a test with the `__thread` annotation in a sample C executable compiled with the Android NDK last year, it would compile, because their lightly patched gcc and clang compilers support it by default, but segfault when run on Android/x86, because the required TLS relocations were not supported by bionic and their dynamic linker.